docs/arch/adr-006-click-for-cli.md

```diff
@@ -0,0 +1,47 @@
+# Click powers dbt-core's CLI
+
+## Context
+ Previous to the implementation of this ADR, the code in `dbt-core` that handled CLI interactions was a combination of [`argparse`](https://docs.python.org/3/library/argparse.html) and a lot of custom code that altered its behavior by subclassing its components. However, as time progressed, it became clear that this approach was lacking in several regards: 
+ * There needed to be a reasonable way to programmatically discover the various CLI commands and their available options. This made integrating other software with the API that the CLI represented especially challenging.
+ * The arguments and argument groups were becoming more extensive and more complex. Because each grouping of arguments was created manually, this led to a lot of duplicated or unneeded code simply for the sake of organization.
+ * `argparse` does not provide any plugin architecture or other methods to add functionality from third parties, leaving us to build any customizations entirely on our own. 
+
+## Decision
+
+ ### Research
+  Research was done to discover an alternative as part of an overall API improvement effort. [Click](https://click.palletsprojects.com/) was selected to replace argparse for the following reasons:
+  * Click provides a more robust ecosystem for building CLIs without being overly heavy-weight.
+  * Click is probably the most widely adapted CLI framework for Python, which provides a significant mindshare in the dev community.
+  * Click's composability allows for future code combining various commands into a single programmatic entry point with very little work.
+  * Click provides a plugin ecosystem that allows us to take advantage of third-party contributions.
```

Comment:
Do you mean things like [click-plugins](https://github.com/click-contrib/click-plugins)? Can we add some example here so people reading this knows better about what are we thinking about.